overlord/state/notices.go

```diff
@@ -211,6 +295,20 @@ type AddNoticeOptions struct {
 
 	// Time, if set, overrides time.Now() as the notice occurrence time.
 	Time time.Time
+
+	// RepeatCheckData is saved for stateful tracking throughout the notice's lifetime.
+	//
+	// NOTES:
+	//	- RepeatCheckData must be JSON marshallable.
+	//	- Setting RepeatCheckData to nil does not remove old state.
+	//	- RepeatCheckData and RepeatCheck cannot be set at the same time.
+	RepeatCheckData interface{}
+
+	// RepeatCheck, if set, returns whether this notice is forced to not be repeated
+	// and also returns the new repeat check data.
+	//
+	// NOTE: Current state can be accessed through oldNotice.GetRepeatCheckValue().
+	RepeatCheck func(oldNotice *Notice, newNoticeOpts *AddNoticeOptions) (repeatOk bool, newRepeatCheckData interface{}, err error)
```

Comment:
All of this strikes me as a rather complex API (which I haven't quite figured out yet). Why do we need both `RepeatCheckData` and the `RepeatCheck` func? How will these be used?

I think it would really help to provide one or two real-world examples of how this will be used (say in the PR description) to see what the intended usage is and help determine whether the complexity is warranted or not.

Reply:
I asked @ZeyadYasser to add also some more realistic unit tests, it will also help clarify some aspects of the new API, atm is not super clear which behavior is essential in it and which accidental or undesired.

Reply:
@benhoyt as suggested by @pedronis, I added a realistic test (i.e. `TestRepeatCheckChangeUpdateScenario`) for the change-update use case.